### DISCUSSION TITLE
What is the guidance around list()?

### DISCUSSION THREAD

**Discussion by cjohnson318** [modular/modular#1300] [Q&A] [answered]
In Python, I usually define empty lists using `list()`, since it screams "list" to me louder than `[]`. Will Mojo support this keyword in future? Is this an anti-pattern in Mojo?

**Comment by soraros** [comment, marked as answer]
Mojo currently lacks a data type equivalent to Python's `list`, which is likely due to ongoing work on traits. An alternative, such as `DynamicVector`, exists, but it is not yet integrated with the `ListLiteral` mechanism. Therefore, relying on constructors (akin to Python's `list()`) remains necessary.

That said, I think `list()` is an anti-pattern in Python. Because `list` is really a dynamic variable that *usually* binds to the `buildins.list`, Python will generate an extra `LOAD_GLOBAL` so the resultant byte code is slightly worse. I don't think it will greatly affect the actual performance of your Python code, it's just a good to know. `dict` has some slight overhead for the same reason.

```python
from dis import dis

def f():
  l = []

def g():
  l = list()

print('f:')
dis(f)
print('g:')
dis(g)
```
Result:
```
f:
  3           0 RESUME                   0

  4           2 BUILD_LIST               0
              4 STORE_FAST               0 (l)
              6 LOAD_CONST               0 (None)
              8 RETURN_VALUE
g:
  6           0 RESUME                   0

  7           2 LOAD_GLOBAL              1 (NULL + list)
             14 PRECALL                  0
             18 CALL                     0
             28 STORE_FAST               0 (l)
             30 LOAD_CONST               0 (None)
             32 RETURN_VALUE
```

Similar overhead won't be observed in Mojo (at least at runtime), as it's compiled and names are resolved statically.

**Reply by cjohnson318** [reply]
Thanks, I think DynamicVector is probably my best bet then.
